### PR TITLE
Feature: SkyblockXPInChat

### PR DESCRIPTION
**Pull request by Thunderblade73**
## Preview
image
## Discord
https://discord.com/channels/997079228510117908/1188397928398716989

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt [hunk 1]
```diff
@@ -1,29 +1,54 @@
...
+    ),
+    riftTime(
+        // language=RegExp
+        "§[a7](?<riftTime>[\\dms ]+)ф.*"
+    ),
+    skyBlockXP(
+        // language=RegExp
+        ".*(§b\\+\\d+ SkyBlock XP §.\\([^()]+\\)§b \\(\\d+/\\d+\\)).*"
     )
+    ;
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt]: can we get the enum names to be all capitals like the rest of the codebase and can you also add a trailing comma to the final one so we get cleaner changelogs in future if a new stat is added

File: src/main/java/at/hannibal2/skyhanni/config/features/chat/ChatConfig.java [hunk 2]
```diff
@@ -118,4 +117,10 @@ public String toString() {
     @ConfigEditorBoolean
     @FeatureToggle
     public boolean translator = false;
+
+    @Expose
+    @ConfigOption(name = "Skyblock XP Chat", desc = "Prints the Skyblock XP message into the chat")
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/chat/ChatConfig.java]: punctuation

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
feature worked and seems very cool, just these two small changes

**Review by CalMWolfs — APPROVED** [review]
lgtm

File: src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt [hunk 3]
```diff
@@ -1,29 +1,60 @@
...
-        if (!LorenzUtils.inSkyBlock) return
-
-        for ((groupName, pattern) in patterns) {
-            pattern.matchMatcher(event.message) {
-                groups[groupName] = group(groupName)
-            }
+enum class ActionBarStatsData(@Language("RegExp") rawPattern: String) { HEALTH(
+    // language=RegExp
+    "§[c6](?<health>[\\d,]+)/[\\d,]+❤.*"
+),
```
Comment (lineargraph) [inline src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt]: Formatting.

File: src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt [hunk 4]
```diff
@@ -1,29 +1,60 @@
...
+
+            entries.mapNotNull { value ->
+                value.pattern.matchMatcher(event.message) {
+                    val newValue = group(1)
+                    if (value.value != newValue) {
+                        value.value = newValue
+                        return@mapNotNull ActionBarValueUpdate(value)
+                    }
+                }
+                null
```
Comment (lineargraph) [inline src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt]: Not sure if this is more clear, but this is something you can do: 

```suggestion
                value.pattern.matchMatcher(event.message) {
                    val newValue = group(1)
                    if (value.value != newValue) {
                        value.value = newValue
                        ActionBarValueUpdate(value)
                    } else null
                }
```

File: src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt [hunk 5]
```diff
@@ -1,29 +1,60 @@
...
+
+        init {
+            entries.forEach { it.pattern }
+        }
+
+        @SubscribeEvent
+        fun onActionBar(event: LorenzActionBarEvent) {
+            if (!LorenzUtils.inSkyBlock) return
+
+            entries.mapNotNull { value ->
```
Comment (lineargraph) [inline src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt]: Not the biggest fan of using `value` as a name here.

**Review by lineargraph — CHANGES_REQUESTED** [review]

**Review by lineargraph — APPROVED** [review]

File: src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt [hunk 6]
```diff
@@ -1,29 +1,60 @@
...
+    SKYBLOCK_XP(
+        // language=RegExp
+        ".*(§b\\+\\d+ SkyBlock XP §.\\([^()]+\\)§b \\(\\d+/\\d+\\)).*"
+    ),
+    ;
+
+    private val repoKey = name.replace("_", ".").lowercase()
+
+    internal val pattern by RepoPattern.pattern("actionbar.$repoKey", rawPattern)
+    var value: String = ""
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/data/ActionBarStatsData.kt:37]: please add `private set`

File: src/main/java/at/hannibal2/skyhanni/config/features/chat/ChatConfig.java [hunk 7]
```diff
@@ -118,4 +117,10 @@ public String toString() {
     @ConfigEditorBoolean
     @FeatureToggle
     public boolean translator = false;
+
+    @Expose
+    @ConfigOption(name = "Skyblock XP Chat", desc = "Prints the Skyblock XP message into the chat.")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/chat/ChatConfig.java]: Please use the wording "send" over "print"

**Review by hannibal002 — CHANGES_REQUESTED** [review]
Does work fine!
image

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**